### PR TITLE
Zadanie ML

### PR DESCRIPTION
**Pull request by Kukann25**

### CODE REVIEW

File: ML_zadanie.ipynb
Comment (nowikarol) [inline ML_zadanie.ipynb:1]: Moje uwagi:
- Do rysowania wykresów używasz zbiorów treningowych.
- Rysujesz dużo wykresów 1D zamiast wykorzystać wyniki z GridSearch do zrobienia wykresów przedstawiających zależności.
- Obecnie wykresy przedstawiają jedną zmienną, a reszta tam jest domyślna.
- Dla parametrów kategorycznych wykres liniowy jest błędem logicznym.

**Review by wsparcie — APPROVED** [review]
bardzo fajna i logiczna argumentacja wyboru recall jako kluczowej metryki (dobra interpretacja false negatives) 
ciekawe i praktyczne uzycie petli do trenowania wielu roznych modeli (od KNN przez RF az po SVC) z gridsearchcv 
czytelen wizualizacje za pomoca macierzy pomylek
przy podziale zbioru (train_test_split) troche brakuje stratify=y do zachowania proporcji klas
skoro we wnioskach to recall jest najwazniejszy, to w gridsearchcv mozna smialo zmienic scoring z accuracy wlasnie na recall
dla modeli bazujacych na obliczaniu dystansu (jak KNN i SVC) w pipelinie brakuje standaryzacji (np. standardscaler), co pewnie podbiloby ich gorsze wyniki

ogolnie bardzo spoko :)

**Review by TsuneKioo — COMMENTED** [review]
Jak już wspomniano wyżej, świetny pomysł z wykorzystaniem pętli do trenowania wszystkich modeli na raz. Wyszedł z tego szybki i kompaktowy trening. 

Bardzo ładne wykresy dla pojedynczych hiperparametrów. 

Faktycznie na początku zabrakło skalowania danych.

**Review by woznin — COMMENTED** [review]
Troche braków z polecenia:
1. Nie mamy żadnych wniosków z poprzedniego zadania, a w poleceniu jasno jest:

> Wypisz obserwacje z pierwszego zadania, które pomogą Ci w tym. Co było przydatne, a co okazało się bezużyteczne?

2. Brak ewaluacji na train i test. Modele liczone są tylko na teście. Bez metryk treningowych nie da się ocenić czy model nadzwyczaj dobrze radzi sobie na treningu i gorzej na tescie (przeuczenie)

3. Pytanie o standaryzację:

> Jak przeprowadzenie standaryzacji danych może wpływać na zachowanie modelu?

 zostaje bez eksperymentu i bez odpowiedzi, mimo że KNN i SVC tego najbardziej potrzebują.


Idąc dalej też troche błędów bardziej metodologicznych:
Bardzo ładnie, ze owinałeś wszystko w test_model, ale dopuszczasz w tej funkcji gigantyczny Leak danych poprzez sprawdzanie najlepszych parametrów na zbiorze testowym - bezposrednio zagladasz do zbioru, który odkładamy na finalny test/traktujemy jakbyśmy go nigdy nie widzieli.  (lepsze wytłumaczenie [tutaj](https://medium.com/@maharishi92vyas/a-data-leakage-mistake-often-made-while-using-gridsearchcv-randomsearchcv-2cd31a2b4591))

>  learning rate dla Gradient Boostingu (recall jako jedyny jest wyższy, ale uważam, że to dobrze bo to najważniejsza matryka - zwiększanie learning rate'a zmniejsza szansę na trafienie w dobre minimum dla gradientu),

Nie zgodze się, accuracy/precision spadają, recall rośnie, bo model zaczyna masowo zgadywać klasę 1. To nie jest „dobry recall”, tylko zdegenerowany klasyfikator. Wybór dużego LR „bo recall jest najważniejszy” jest po prostu błędem. 
Dodatkowo, nawet jeśli już patrzymy na recall jako na najważniejszą metrykę to dlaczego grid search patrzy na 

>         scoring="accuracy",

I takie mniejsze komentarze:
Jak mamy takie 'stałe' bloki kodu, warto je wyrzucić do jakiegoś pliku obok w stylu utils.py czy coś, żeby nie mieć w raporcie takiej ścianki kodu 😉 

Ogólnie notebook mi się podoba, ale popraw proszę te błędy:
- data leak
- brakujace wnioski z poprzedniego zadania
- podglad testów train vs test
- standaryzacja vs jej brak
- i oczywiscie scoring (najbardziej zwróć uwagę, czy dając właśnie np. recall nie będziemy zakłamywać jakości modelu - wybierzemy coś co radzi sobie gorzej od innych pod każdym względem oprócz wybranej metryki).
